core/shared/src/main/scala/scalaz/zio/ExitResult.scala

```diff
@@ -9,56 +11,192 @@ package scalaz.zio
     self match {
       case Completed(v)  => completed(v)
-      case Failed(e, ts) => failed(e, ts)
-      case Terminated(e) => interrupted(e)
+      case Failed(cause) => terminated(cause)
     }
+}
+object ExitResult {
+
+  final case class Completed[A](value: A)     extends ExitResult[Nothing, A]
+  final case class Failed[E](cause: Cause[E]) extends ExitResult[E, Nothing]
+
+  final def checked[E](error: E): ExitResult[E, Nothing] = ExitResult.Failed(Cause.checked(error))
+  final def interrupted(causes: List[Throwable]): ExitResult[Nothing, Nothing] =
+    ExitResult.Failed(Cause.interrupt(causes))
+  final def unchecked(t: Throwable): ExitResult[Nothing, Nothing] = ExitResult.Failed(Cause.unchecked(t))
+
+  sealed abstract class Cause[+E] extends Product with Serializable { self =>
+    import Cause._
+    final def ++[E1 >: E](that: Cause[E1]): Cause[E1] =
+      Then(self, that)
+
+    final def &&[E1 >: E](that: Cause[E1]): Cause[E1] =
+      Both(self, that)
+
+    final def map[E1](f: E => E1): Cause[E1] = self match {
+      case Checked(value)      => Checked(f(value))
+      case c @ Unchecked(_)    => c
+      case c @ Interruption(_) => c
 
-  final def zipWith[E1 >: E, B, C](that: ExitResult[E1, B])(f: (A, B) => C): ExitResult[E1, C] = (self, that) match {
-    case (ExitResult.Completed(a), ExitResult.Completed(b))   => ExitResult.Completed(f(a, b))
-    case (ExitResult.Failed(e, ts), rb)                       => ExitResult.Failed(e, combine(ts, rb))
-    case (ExitResult.Terminated(ts), rb)                      => ExitResult.Terminated(combine(ts, rb))
-    case (ExitResult.Completed(_), ExitResult.Failed(e, ts))  => ExitResult.Failed(e, ts)
-    case (ExitResult.Completed(_), ExitResult.Terminated(ts)) => ExitResult.Terminated(ts)
+      case Then(left, right) => Then(left.map(f), right.map(f))
+      case Both(left, right) => Both(left.map(f), right.map(f))
+    }
+
+    final def checked[E1 >: E]: List[E1] =
+      self
+        .fold(List.empty[E1]) {
+          case (z, Checked(v)) => v :: z
+        }
+        .reverse
+
+    final def unchecked: List[Throwable] =
+      self
+        .fold(List.empty[Throwable]) {
+          case (z, Unchecked(v)) => v :: z
+        }
+        .reverse
+
+    final def interruptions: Option[List[Throwable]] =
+      self
+        .fold(Option.empty[List[Throwable]]) {
+          case (z, Interruption(v)) =>
+            val ts = v.fold[List[Throwable]](Nil)(List(_))
+            z.map(ts ++ _) orElse Some(ts)
+        }
+        .map(_.reverse)
+
+    final def fold[Z](z: Z)(f: PartialFunction[(Z, Cause[E]), Z]): Z =
+      (f.lift(z -> self).getOrElse(z), self) match {
+        case (z, Then(left, right)) => right.fold(left.fold(z)(f))(f)
+        case (z, Both(left, right)) => right.fold(left.fold(z)(f))(f)
+
+        case (z, _) => z
+      }
+
+    final def toThrowable: Throwable = FiberFailure(self)
+
+    final def toIO: IO[E, Nothing] = IO.fail0(self)
```

Comment:
This should be deleted. It's so simple now it's just proxying and adding one unnecessary layer of indirection.